### Exercício 1

<small>

Inspecionando um Array NumPy

Um objeto `numpy` possui atributos e métodos, herdados da classe `numpy.ndarray`. Crie um array e inspecione os seguintes atributos:

- **shape**: uma tupla que contém o número de elementos (tamanho) em cada dimensão do array;
- **size**: número total de elementos do array;
- **ndim**: número de dimensões (eixos);
- **nbytes**: número de bytes utilizado no armazenamento dos dados;
- **dtype**: tipo de dado dos elementos do array.
</small>

In [1]:
import numpy as np

In [ ]:
array = np.array(range(1000))

In [28]:
print('Shape: ', array.shape)
print('Size: ',array.size)
print('Números de dimensões: ', array.ndim)
print('Número de Bytes: ', array.nbytes)
print('Tipo dos elementos do array: ', array.dtype)

Shape:  (1000,)
Size:  1000
Números de dimensões:  1
Número de Bytes:  8000
Tipo dos elementos do array:  int64


### Exercício 2

<small>

Com base no exercício anterior crie uma função que receba um array e mostre na tela seus atributos. Verifique se o argumento inserido é de fato um objeto numpy. Use também a função built-in type() para verificar a classe do objeto.

<small>

In [16]:
def inspecionar_array(array):
    if(isinstance(array,np.ndarray)):
        print('Shape: ', array.shape)
        print('Size: ',array.size)
        print('Números de dimensões: ', array.ndim)
        print('Número de Bytes: ', array.nbytes)
        print('Tipo dos elementos do array: ', array.dtype)
        print('Tipo do array: ', type(array))
        print('Array: ', array)
    else:
        print('O objeto inserido não é um array NumPy.')

In [17]:
inspecionar_array(array)

Shape:  (1000,)
Size:  1000
Números de dimensões:  1
Número de Bytes:  8000
Tipo dos elementos do array:  int64
Tipo do array:  <class 'numpy.ndarray'>
Array:  [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53
  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 

In [20]:
list_ = [1, 2, 3, 4]
print(type(list_))
inspecionar_array(list_)

<class 'list'>
O objeto inserido não é um array NumPy.


### Exercício 3

<small>
Use a função np.array() para criar os arrays unidimensionais listados abaixo:

a) Use uma lista como argumento e obtenha:

    array([-1,  4,  7,  9])

b) Use a função range() e obtenha:

    array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

c)Use uma tupla como argumento e obtenha:
    
    array([10, 11, 12, 13, 14])

d) Use a função range() e obtenha:
    
    array([1990, 1995, 2000, 2005, 2010, 2015, 2020])

e) Use a função range() e obtenha:

    array([2020, 2010, 2000, 1990, 1980, 1970, 1960, 1950, 1940, 1930, 1920,
       1910, 1900])

f) Obtenha:

    array(100)

<small>

In [35]:
list_ = [1, 2, 3, 4]
print(type(list_))
list_ = np.array(list_)
print(type(list_))
print(list_)

<class 'list'>
<class 'numpy.ndarray'>
[1 2 3 4]


In [30]:
array = np.array(range(11))
print(array)
print(type(array))

[ 0  1  2  3  4  5  6  7  8  9 10]
<class 'numpy.ndarray'>


In [36]:
tupla = (10,20,30,40,50)
print(type(tupla))
tupla = np.array(tupla)
print(type(tupla))
print(tupla)

<class 'tuple'>
<class 'numpy.ndarray'>
[10 20 30 40 50]


In [38]:
array = np.array(range(1990,2025,5))
print(array)
print(type(array))

[1990 1995 2000 2005 2010 2015 2020]
<class 'numpy.ndarray'>


In [42]:
array = np.array(range(2020,1890,-10))
print(array)
print(type(array))

[2020 2010 2000 1990 1980 1970 1960 1950 1940 1930 1920 1910 1900]
<class 'numpy.ndarray'>


In [45]:
num = 100
print(type(num))
num = np.array(num)
print(type(num))
print(num)

<class 'int'>
<class 'numpy.ndarray'>
100


### Exercício 4

<small>
A função np.ones() permite criar um array preenchido com o número 1 em todas as suas entradas. Podemos especificar como argumentos o shape e o tipo de dado dtype do array. Use esta função para criar os arrays a seguir:

a)
    array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])


b)
    array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

c)
    array([[1, 1],
       [1, 1],
       [1, 1]])


<small>

In [50]:
arrayOne = np.ones(11)
print(arrayOne)

[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]


In [54]:
arrayOne = np.ones((4,4))
print(arrayOne)

[[1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]]


In [59]:
arrayOne = np.ones((3,2),dtype=int)
print(arrayOne)

[[1 1]
 [1 1]
 [1 1]]


### Exercício 5

<small>
A função np.zeros() permite criar um array preenchido com o número 0 em todas as suas entradas. Podemos especificar como argumentos o shape e o tipo de dado dtype do array (default é o float). Use esta função para criar os arrays a seguir:

a)
    array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

b)
    array([[0., 0.],
       [0., 0.]])  
       
c)
    array([[0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.]])

<small>

In [57]:
arrayZeros = np.zeros(11)
print(arrayZeros)

[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]


In [58]:
arrayZeros = np.zeros((2,2))
print(arrayZeros)

[[0. 0.]
 [0. 0.]]


In [60]:
arrayZeros = np.zeros((4,6))
print(arrayZeros)

[[0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]]


### Exercício 6

<small>
Crie os arrays a seguir:

a)
       array([[10],
       [15],
       [20],
       [25],
       [30],
       [35],
       [40],
       [45]])

b)

array([-50, -45, -40, -35, -30, -25, -20, -15, -10,  -5])

c)

array([[ 50,  52,  54,  56,  58,  60,  62,  64,  66,  68],
       [ 70,  72,  74,  76,  78,  80,  82,  84,  86,  88],
       [ 90,  92,  94,  96,  98, 100, 102, 104, 106, 108],
       [110, 112, 114, 116, 118, 120, 122, 124, 126, 128],
       [130, 132, 134, 136, 138, 140, 142, 144, 146, 148]])
       

Dica: use np.arange() ou np.array(range()).
<small>

In [65]:
array = np.arange(10,50,5).reshape(-1,1)
print(array)

[[10]
 [15]
 [20]
 [25]
 [30]
 [35]
 [40]
 [45]]


In [64]:
array = np.array(range(-50,0,5))
print(array)

[-50 -45 -40 -35 -30 -25 -20 -15 -10  -5]


In [67]:
array = np.arange(50,150,2).reshape((5,10))
print(array)

[[ 50  52  54  56  58  60  62  64  66  68]
 [ 70  72  74  76  78  80  82  84  86  88]
 [ 90  92  94  96  98 100 102 104 106 108]
 [110 112 114 116 118 120 122 124 126 128]
 [130 132 134 136 138 140 142 144 146 148]]


### Exercício 7

<small>
Utilize a função np.full(shape, fill_value, dtype) para criar os arrays a seguir. Esta função permite criar um array com um shape e dtype definidos pelo usuário, preenchido com um determinado valor.

a)
    array([5., 5., 5., 5., 5., 5., 5., 5., 5., 5.])

b)

    array([[-5., -5., -5., -5.],
       [-5., -5., -5., -5.],
       [-5., -5., -5., -5.],
       [-5., -5., -5., -5.]])

c)
    array([[3, 3, 3],
       [3, 3, 3],
       [3, 3, 3]])

d)

    array([[50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50]])

Dica: outra possibilidade seria obter os arrays com uma operação diferente, a partir da multiplicação de uma matriz preenchida com número 1 por um determinado valor.
<small>

In [70]:
arrayFull = np.full((1,10),5,float)
print(arrayFull)

[[5. 5. 5. 5. 5. 5. 5. 5. 5. 5.]]


In [76]:
arrayFull = np.full((3,3),3,int)
print(arrayFull)

[[3 3 3]
 [3 3 3]
 [3 3 3]]


In [ ]:
arrayFull = np.full((4,4),-5,float)
print(arrayFull)

In [74]:
arrayFull = np.full((10,10),50,int)
print(arrayFull)

[[50 50 50 50 50 50 50 50 50 50]
 [50 50 50 50 50 50 50 50 50 50]
 [50 50 50 50 50 50 50 50 50 50]
 [50 50 50 50 50 50 50 50 50 50]
 [50 50 50 50 50 50 50 50 50 50]
 [50 50 50 50 50 50 50 50 50 50]
 [50 50 50 50 50 50 50 50 50 50]
 [50 50 50 50 50 50 50 50 50 50]
 [50 50 50 50 50 50 50 50 50 50]
 [50 50 50 50 50 50 50 50 50 50]]


### Exercício 8

<small>
    Defina o array K e verifique se há dados ausentes (missing data). Use a função np.isnan().

array([[ nan,   1.,  10.,   0.],
       [ nan,  nan,  nan,  nan],
       [100.,  50.,  nan, -25.],
       [ 30.,  nan,  nan, 130.]])
<small>

In [81]:
K = np.array([
    [np.nan, 1, 10, 0],
    [np.nan,np.nan,np.nan,np.nan],
    [100, 50, np.nan, -25],
    [30, np.nan, np.nan, 130]])

In [83]:
Knan = np.isnan(K)

In [84]:
Knan

array([[ True, False, False, False],
       [ True,  True,  True,  True],
       [False, False,  True, False],
       [False,  True,  True, False]])

In [85]:
print(np.isnan(K))
print(f'Há {np.isnan(K).sum()} dados ausentes em K.')

[[ True False False False]
 [ True  True  True  True]
 [False False  True False]
 [False  True  True False]]
Há 8 dados ausentes em K.


### Exercício 9

<small>
A função np.linspace() retorna um array com números contendo espaçamento uniforme em um intervalo definido. Crie um array com números dentro do intervalo [1,21] em 5 partes.

Retorno esperado:

array([ 1.,  6., 11., 16., 21.])
<small>

In [87]:
array = np.linspace(1,21,5)
print(array)

[ 1.  6. 11. 16. 21.]


### Exercício 10

<small>
A função np.logspace() retorna números com espaçamento uniforme em uma escala logarítmica. Crie um array com números dentro do intervalo [1, 51] em 10 partes com a função np.logspace().

Retorno esperado:

array([[1.00000000e+01],
       [3.59381366e+06],
       [1.29154967e+12],
       [4.64158883e+17],
       [1.66810054e+23],
       [5.99484250e+28],
       [2.15443469e+34],
       [7.74263683e+39],
       [2.78255940e+45],
       [1.00000000e+51]])
<small>

In [90]:
array = np.logspace(1,51,10).reshape(-1,1)
print(array)

[[1.00000000e+01]
 [3.59381366e+06]
 [1.29154967e+12]
 [4.64158883e+17]
 [1.66810054e+23]
 [5.99484250e+28]
 [2.15443469e+34]
 [7.74263683e+39]
 [2.78255940e+45]
 [1.00000000e+51]]


### Exercício 11

<small>
Use a função np.diag() para construir o array a seguir:

array([[ 5,  0,  0,  0,  0,  0],
       [ 0, 10,  0,  0,  0,  0],
       [ 0,  0, 15,  0,  0,  0],
       [ 0,  0,  0, 20,  0,  0],
       [ 0,  0,  0,  0, 25,  0],
       [ 0,  0,  0,  0,  0, 30]])
<small>

In [93]:
arrayDig = np.diag(range(5,35,5))
print(arrayDig)

[[ 5  0  0  0  0  0]
 [ 0 10  0  0  0  0]
 [ 0  0 15  0  0  0]
 [ 0  0  0 20  0  0]
 [ 0  0  0  0 25  0]
 [ 0  0  0  0  0 30]]


### Exercício 12

<small>
Use a função np.eye() (ou ainda np.identity()) para obter o array seguir.

array([[1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 1., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 1., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 1., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 1., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 1.]])


Dica: essa matriz é definida como matriz identidade, neste caso de ordem 10, e assume grande relevância em diversos cálculos computacionais, como veremos mais adiante.
<small>

In [98]:
arrayID = np.eye(10)
print(arrayID)

[[1. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 1. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 1. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 1. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 1. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 1. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 1. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]]


### Exercício 14

<small>
Obtenha a soma dos elementos de cada array a seguir. Use a função np.sum(). Implemente uma solução com loop for, para tanto armazene os vetores em uma lista.

import numpy as np
array1 = np.arange(1,11)
array2 = np.linspace(1,50,10)
array3 = np.array(range(51))
array4 = np.logspace(1,10,5)
array5 = np.diag(range(5,31,5))

Retorno esperado:

Soma dos elementos do array 1:  55
Soma dos elementos do array 2:  255.0
Soma dos elementos do array 3:  1275
Soma dos elementos do array 4:  10056552148.564463
Soma dos elementos do array 5:  105
<small>

In [ ]:
array1 = np.arange(1,11)
array2 = np.linspace(1,50,10)
array3 = np.array(range(51))
array4 = np.logspace(1,10,5)
array5 = np.diag(range(5,31,5))

arrayGeral = [array1,array2,array3,array4,array5]

In [107]:
it = 1
for array in arrayGeral:
    print('Soma dos elementos do array',it,': ',array.sum())
    it = it + 1

Soma dos elementos do array 1 :  55
Soma dos elementos do array 2 :  255.0
Soma dos elementos do array 3 :  1275
Soma dos elementos do array 4 :  10056552148.564463
Soma dos elementos do array 5 :  105


In [108]:
arrays = [array1, array2, array3, array4, array5]
for index, array in enumerate(arrays, start = 1):
    print(f'Soma dos elementos do array {index}: ', array.sum())

Soma dos elementos do array 1:  55
Soma dos elementos do array 2:  255.0
Soma dos elementos do array 3:  1275
Soma dos elementos do array 4:  10056552148.564463
Soma dos elementos do array 5:  105


### Exercício 15

<small>
A partir da matriz definida abaixo obtenha: a) soma total dos elementos; b) soma das linhas ; c) soma das colunas.

import numpy as np
matriz = np.array(np.random.randint(1,100,15)).reshape(5,3)
<small>

In [110]:
matriz = np.array(np.random.randint(1,100,15)).reshape(5,3)
print(matriz)

[[80 57 50]
 [24 32 39]
 [ 3 37  2]
 [81 24 76]
 [89 13 96]]


In [113]:
print('Soma total dos elementos: ', matriz.sum())
print('Soma das linhas: ', matriz.sum(axis = 1))
print('Soma das linhas: ', matriz.sum(axis = 0))

Soma total dos elementos:  703
Soma das linhas:  [187  95  42 181 198]
Soma das linhas:  [277 163 263]


### Exercício 16

<small>
Obtenha o valor máximo e o valor mínimo de cada array abaixo. Implemente uma solução com loop for.

import numpy as np
array1 = np.array([-1, 10, 3, 4, 7, 27])
array2 = np.random.randint(-100,300,30)
array3 = np.random.randn(10)
array4 = np.array(range(50,10,-2))

Dica: para obter o valor máximo e mínimo de um array podemos usar array.max() e array.min() ( ou ainda max(array) e min(array)).
<small>

In [114]:
array1 = np.array([-1, 10, 3, 4, 7, 27])
array2 = np.random.randint(-100,300,30)
array3 = np.random.randn(10)
array4 = np.array(range(50,10,-2))

In [116]:
arrays = [array1, array2, array3, array4, array5]
for index, array in enumerate(arrays, start = 1):
    print(f'O maior valor do array {index}: ', array.max())
    print(f'O valor mínimo do array {index}: ', array.min())

O maior valor do array 1:  27
O valor mínimo do array 1:  -1
O maior valor do array 2:  289
O valor mínimo do array 2:  -92
O maior valor do array 3:  2.3517447124830775
O valor mínimo do array 3:  -1.7081997612873503
O maior valor do array 4:  50
O valor mínimo do array 4:  12
O maior valor do array 5:  30
O valor mínimo do array 5:  0


In [117]:
for index, array in enumerate(arrays, start = 1):
    print(f'Array {index}')
    print(f'Valor máximo: ', array.max())
    print(f'Valor mínimo: ', array.min())
    print()

Array 1
Valor máximo:  27
Valor mínimo:  -1

Array 2
Valor máximo:  289
Valor mínimo:  -92

Array 3
Valor máximo:  2.3517447124830775
Valor mínimo:  -1.7081997612873503

Array 4
Valor máximo:  50
Valor mínimo:  12

Array 5
Valor máximo:  30
Valor mínimo:  0



### Exercício 17

<small>
Crie uma função que receba uma matriz como entrada e calcule:

    a) Valor máximo da matriz;

    b) Valor máximo por coluna;

    c) Valor máximo por linha;

    d) Valor mínimo da matriz;

    e) Valor mínimo por coluna;

    f) Valor mínimo por linha.



Exemplo de retorno para a matriz definida abaixo:

Input:

matriz = np.random.randint(-1000,1000,50).reshape(10,5)
info_matriz(matriz)
Output:

Valor máximo da matriz:  987
Valor máximo por coluna:  [959 987 597 748 719]
Valor máximo por linha:  [ 987  312  982  204  719  689  101  793 -460  959]
Valor mínimo da matriz:  -969
Valor mínimo por coluna:  [-878 -966 -969 -884 -881]
Valor mínimo por linha:  [  30 -878 -335 -863  -40 -289 -660 -239 -969 -966]
<small>

In [118]:
matriz = np.random.randint(-1000,1000,50).reshape(10,5)
matriz

array([[-948, -179,  600, -241,  431],
       [-994,  944, -558,  944,  114],
       [-986,  421,  831,  -66,  332],
       [ 364,  145,  942, -633, -561],
       [-388,    4,  230,   97, -690],
       [-206, -316,  536,  222, -408],
       [-676,   55,  816, -883,  -85],
       [ 358,  639,  192,   92, -679],
       [-179,  143, -779, -223,  281],
       [ 569,  386,  239,  171,  530]], dtype=int32)

In [119]:
def info_matriz(matriz):
    print('Valor máximo da matriz: ', matriz.max())
    print('Valor máximo por coluna: ', matriz.max(axis=0))
    print('Valor máximo por linha: ', matriz.max(axis=1))
    print('Valor mínimo da matriz: ', matriz.min())
    print('Valor mínimo por coluna: ', matriz.min(axis=0))
    print('Valor mínimo por linha: ', matriz.min(axis=1))

In [120]:
info_matriz(matriz)

Valor máximo da matriz:  944
Valor máximo por coluna:  [569 944 942 944 530]
Valor máximo por linha:  [600 944 831 942 230 536 816 639 281 569]
Valor mínimo da matriz:  -994
Valor mínimo por coluna:  [-994 -316 -779 -883 -690]
Valor mínimo por linha:  [-948 -994 -986 -633 -690 -408 -883 -679 -779  171]


### Exercício 18

<small>
Com base no array definido abaixo obtenha sua média, mediana, variância e desvio padrão.
<small>

In [122]:
arr = np.random.normal(40,0.5,size = (300))
print(arr)

[39.83363512 39.41944226 39.44808665 39.79401648 40.26294106 40.03486649
 40.40403563 39.53547903 40.65096671 39.87682506 40.99194045 40.39318789
 40.30212698 40.41245106 39.54599507 39.38953261 40.07838893 38.71925385
 41.35113373 40.99233608 39.36714515 40.23186737 39.33069795 39.3640436
 40.19992511 40.68120988 40.75731992 39.99835847 39.636795   40.42856425
 40.28708221 39.33239542 40.00126251 39.40928727 40.44918185 39.96481528
 41.30585158 39.1939162  39.68839876 40.1743051  39.72905594 40.08766756
 40.02449563 40.56118298 39.60615105 40.07442957 39.83680582 39.2710817
 39.17153029 39.6954738  39.5022693  39.86652298 40.02426237 40.35351015
 39.66932203 40.78750052 39.96160135 40.3433981  40.3016303  39.84050387
 40.03376228 39.59368966 40.39959237 39.75007084 39.47380098 39.37090691
 40.69191125 40.24057899 40.37770026 39.63919998 39.94138095 39.36087049
 40.3074202  40.79613777 39.1009936  39.63397685 39.87027292 40.01015743
 40.02767629 39.1689758  39.69276399 40.42802782 39.1

In [126]:
print('média: ', arr.mean())
print(f'Mediana: {np.median(arr)}')
print('desvio padrão: ', arr.std())
print('variância: ', arr.var())

média:  40.02218007322433
Mediana: 40.02420910024381
desvio padrão:  0.5081707749550215
variância:  0.2582375365183871


### Exercício 19

<small>
Com base nos arrays definidos a seguir obtenha a matriz de covariância e a matriz de correlação.
<small>

In [127]:
x = np.random.rand(100)
y = np.random.uniform(size=100)

In [130]:
print(f'Matriz de correlação: {np.corrcoef(x,y)}')
print()
print(f'Matriz de covariância: {np.cov(x,y)}')

Matriz de correlação: [[ 1.         -0.07800824]
 [-0.07800824  1.        ]]

Matriz de covariância: [[ 0.09028271 -0.00743186]
 [-0.00743186  0.10053323]]


### Exercício 20

<small>
Obtenha o índice do valor máximo do array A.
<small>

In [2]:
A = np.array([[11, 15, 33, 105],
              [1, 140, 45, 90],
              [67, 230, 78, 99]])

In [3]:
print(A.argmax())

9


In [4]:
print(f'Valor máximo do array A: {A.max()}')
print(f'Índice do valor máximo do array A: {A.argmax()}')

Valor máximo do array A: 230
Índice do valor máximo do array A: 9


### Exercício 21

<small>
Obtenha o índice do valor mínimo do array B.
<small>

In [5]:
B = np.array([
    [73, 22, 20, 62,  3],
    [68, 84, 27, 59, 93],
    [48, 52,  4, 55, 50],
    [42, 23, 63, 67,  4],
    [44, 30, 41, 31, 29]])


In [7]:
print(f'Valor mínimo do array B: {B.min()}')
print(f'Índice do valor mínimo do array B: {B.argmin()}')

Valor mínimo do array B: 3
Índice do valor mínimo do array B: 4


### Exercício 22

<small>
Escreva uma função que receba um array NumPy e mostre na tela um conjunto de  estatísticas descritivas deste objeto.
<small>

In [8]:
array = np.array([-57, 101, 270, 130, 144])
print(array)

[-57 101 270 130 144]


In [9]:
def info_array(array):
    print('Valor máximo: ',array.max())
    print('Valor mínimo: ',array.min())
    print('Média: ',np.median(array))
    print('Mediana: ', array.mean())
    print('Variância: ', array.var())
    print('Desvio padrão: ', array.std())

In [10]:
info_array(array)

Valor máximo:  270
Valor mínimo:  -57
Média:  130.0
Mediana:  117.6
Variância:  10967.439999999999
Desvio padrão:  104.72554607162475


In [11]:
def info_array(array):
        
    if isinstance(array,np.ndarray):
        print(f'Valor máximo: {np.max(array)}')
        print(f'Valor mínimo: {np.min(array)}')
        print(f'Média: {np.mean(array)}')
        print(f'Mediana: {np.median(array)}')
        print(f'Variância: {np.var(array)}')
        print(f'Desvio padrão: {np.std(array)}')
    else:
        return None
    
if __name__ == '__main__':
    array = np.array([-57, 101, 270, 130, 144])
    info_array(array)

Valor máximo: 270
Valor mínimo: -57
Média: 117.6
Mediana: 130.0
Variância: 10967.439999999999
Desvio padrão: 104.72554607162475


### Exercício 23

<small>
Use a função np.unique() para obter um array com os elementos únicos do array abaixo. Para tanto, armazene o novo array em uma variável.
<small>

In [12]:
A = np.array([
    [1, 3, 3, 4, 5],
    [4, 7, 11, 3, 5],
    [1, 11, 4, 4, 11]])

In [13]:
arr_unique = np.unique(A)
print(arr_unique)

[ 1  3  4  5  7 11]


### Exercício 24

<small>
Indexação de arrays 1-D
Considerando os arrays abaixo:
Obtenha:

a) o primeiro, o terceiro e o último elemento de cada array;

b) inverta o array;

c) pegue o número 20 em ambos os arrays e armazene a soma em uma variável.
<small>

In [14]:
array1 = np.array([10, 20, 30, 40])
array2 = np.array([100, 30, 5, 1, 40, 100, 20, 130, 155,170])

In [16]:
print(array1[0],array1[2],array1[-1])
print(array1[::-1])
print(array2[0],array2[2],array2[-1])
print(array2[::-1])
soma = array1[1]+array2[6]
print(soma)

10 30 40
[40 30 20 10]
100 5 170
[170 155 130  20 100  40   1   5  30 100]
40


In [18]:
arrays = [array1, array2]
for index, array in enumerate(arrays, start = 1):
    print(f'Array {index}')
    print(f'Primeiro elemento: {array[0]}')
    print(f'Terceiro elemento: {array[2]}')
    print(f'Último elemento: {array[-1]}')
    print(f'Array invertido: {array[::-1]}')
 
soma = array1[1]+array2[6]  #ou: array1[-3]+array2[-4]

Array 1
Primeiro elemento: 10
Terceiro elemento: 30
Último elemento: 40
Array invertido: [40 30 20 10]
Array 2
Primeiro elemento: 100
Terceiro elemento: 5
Último elemento: 170
Array invertido: [170 155 130  20 100  40   1   5  30 100]


### Exercício 25

<small>
Acesse os números 9 e 90 do array abaixo.
<small>

In [19]:
ar = np.array([[-1, 33, 44, 9, 1],[11, 1, 0, 90, 44]])

In [20]:
print(ar[0][3],ar[1][3])

9 90


### Exercício 26

<small>
Acesse os números 5, 9, 3 e 21 do array abaixo.
<small>

In [21]:
ar = np.array([
    [[1, 1, 5, 1],
     [1, 1, 1, 9]],
    [[1, 3, 1, 1],
     [1, 1, 21, 1]]
              ])

In [27]:
print(ar[0][0][2])
print(ar[0][1][3])
print(ar[1][0][1])
print(ar[1][1][2])

5
9
3
21


### Exercício 27

<small>
Com base no array abaixo obtenha o slice do índice 2 até o índice 6.
<small>

In [28]:
ar = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])

In [30]:
ar = ar[2:7]
print(ar)

[5 6]
